packages/notifications/src/error.ts

```diff
@@ -0,0 +1,87 @@
+import { AnyPromiseFn } from '@terra-ui-packages/core-utils';
+
+import { NotificationsContract, Notifier } from './useNotifications';
+
+export type ErrorCallback = (error: unknown) => void | Promise<void>;
+
+/**
+ * Should return true of a given error should be ignored
+ */
+export type IgnoreErrorDecider = (title: string, obj?: unknown) => boolean;
+
+/**
+ * Invoke the `callback` with any error thrown when evaluating the async `fn` with `...args`.
+ */
+export const withErrorHandling = (callback: ErrorCallback) => {
+  return <F extends AnyPromiseFn>(fn: F) => {
+    return (async (...args: Parameters<F>) => {
+      try {
+        return await fn(...args);
+      } catch (error) {
+        await callback(error);
+      }
+    }) as F;
+  };
+};
+
+/**
+ * Return a Promise to the result of evaluating the async `fn` with `...args` or undefined if
+ * evaluation fails.
+ */
+export const withErrorIgnoring = withErrorHandling(() => {});
+
+/**
+ * allows composition of NotificationsContract dependency when useful for code reuse
+ * @param notifier
+ */
+export const makeNotificationsContract = (
+  notifier: Notifier,
+  ignoreError: IgnoreErrorDecider
+): NotificationsContract => ({
```

Comment:
Can we turn the second parameter into an options object to clarify its purpose / make it optional?

```suggestion
interface NotificationsContractOptions {
  ignoreError?: IgnoreErrorDecider;
}

export const makeNotificationsContract = (
  notifier: Notifier,
  options: NotificationsContractOptions = {}
): NotificationsContract => {
  const { ignoreError = () => false } = options;
```

Reply:
Also, maybe `shouldIgnoreError`? I think a `should...` prefix is a common convention for this sort of function (React's `shouldComponentUpdate`, etc).

Reply:
I think my proposal would be an `ErrorReporting` interface and a function `makeNotificationsErrorReporting` or something like that. Then the `NotificationsContract` returned by `useNotifications` could be a union type of `Notifier` and `ErrorReporting`.

Reply:
right, but things are a bit entangled.... reportError method (which is the key functionality used by all the methods), needs to call .notify()... hence why it's all constructed together in the proposed implementation of makeNotificationsProvider

Reply:
Yeah, I think this comment was resolved by consolidating everything into `NotificationsProvider` and thus avoiding "error reporting" as a thing.